### PR TITLE
Jwt setup

### PR DESCRIPTION
**Pull request by Devyani606**

### CODE REVIEW

File: app/controllers/v1/base_controller.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    before_action :authenticate!
+
+    attr_reader :current_tenant, :current_user
+
+    def render_json(message, data, status_code=:ok)
+      render json: {message: message, data: data}, status: status_code
+    end
+
+    def authenticate!
+      # return invalid_authentication if !payload || !JsonWebToken.valid_payload(payload)
```
Comment (smriti154) [inline app/controllers/v1/base_controller.rb]: Remove the commented code

File: app/controllers/v1/base_controller.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      invalid_authentication unless @current_user
+    end
+
+    def validate_jwt_token!
+      if request.headers["Authorization"].present?
+        token = request.headers["Authorization"]
+        @payload = JsonWebToken.decode(token)
+        return token_invalid unless JsonWebToken.valid_payload(@payload)
+      else
+        # render json: {message: "Header Not Present"}, status: :unauthorized
```
Comment (smriti154) [inline app/controllers/v1/base_controller.rb]: Remove the commented code

File: app/controllers/v1/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+module V1
+  class SessionsController < BaseController
+    skip_before_action :authenticate!, only: :login
+
+    def login
+      @user = User.find_by_email(params[:login][:email])
+      if @user&.authenticate(params[:login][:password])
+        token = JsonWebToken.encode(user_id: @user.id, slug: "hdfc")
```
Comment (smriti154) [inline app/controllers/v1/sessions_controller.rb]: slug will be hard-coded?

File: app/controllers/v1/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        render_json("User Login Succesfully", auth_token: token)
+      else
+        render_json("Unauthorized User", "User is Not Present", :unauthorized)
+      end
+    end
+
+    def decode_token
+      puts(User.find(1))
+      puts("current", current_user.inspect, current_tenant.inspect)
+    end
```
Comment (smriti154) [inline app/controllers/v1/sessions_controller.rb]: What is the purpose of this method?

File: app/controllers/v1/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+    def login_params
+      params.require(:login).permit(:email, :password)
```
Comment (smriti154) [inline app/controllers/v1/sessions_controller.rb]: slug will be passed as params

File: db/migrate/20200203105647_create_users.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+  def change
+    create_table :users do |t|
+      t.string :name
+      t.string :email
+      t.string :password_digest
+
+      t.timestamps
+    end
+  end
+end
```
Comment (smriti154) [inline db/migrate/20200203105647_create_users.rb]: Why you need this?

File: spec/requests/v1/authentication_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+describe V1::AuthenticationController do
+end
```
Comment (smriti154) [inline spec/requests/v1/authentication_controller_spec.rb:6]: there is no controller with name Authentication controller

File: public/docs/v1/index.html [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    <div id="container">
+      <div id="operations">
+        <h3>API Operations</h3>
+      </div>
+      <div id="content">
+        <h1>Documentation for v1</h1>
+      </div>
+    </div>
+  </body>
+</html>
```
Comment (smriti154) [inline public/docs/v1/index.html]: Remove this

File: public/docs/v1/style.css [hunk 9]
```diff
@@ -0,0 +1,5 @@
+body {margin: 0; background-color: #fff; color: #000; font-family: Arial,sans-serif;}
+content {margin-left: 200px;}
+content h1 {text-align: center;}
+operations {float: left; width: 200px; border-right: 1px solid #ccc;}
+operations h3 {text-align: center;}
```
Comment (smriti154) [inline public/docs/v1/style.css]: Remove this

**Comment by smriti154** [comment]
@Devyani606 There are conflicts in your branch, update your branch with development

File: app/controllers/v1/base_controller.rb (hunk 1 above)
Comment (Devyani606) [reply app/controllers/v1/base_controller.rb]: done

File: app/controllers/v1/sessions_controller.rb (hunk 3 above)
Comment (Devyani606) [reply app/controllers/v1/sessions_controller.rb]: done

File: spec/requests/v1/authentication_controller_spec.rb (hunk 7 above)
Comment (Devyani606) [reply spec/requests/v1/authentication_controller_spec.rb:6]: done

File: public/docs/v1/index.html (hunk 8 above)
Comment (Devyani606) [reply public/docs/v1/index.html]: done

File: app/controllers/v1/base_controller.rb (hunk 2 above)
Comment (Devyani606) [reply app/controllers/v1/base_controller.rb]: done

File: app/controllers/v1/sessions_controller.rb (hunk 4 above)
Comment (Devyani606) [reply app/controllers/v1/sessions_controller.rb]: These method is especially written for the fetching the current user and current tenant from the base controller

File: db/migrate/20200203105647_create_users.rb (hunk 6 above)
Comment (Devyani606) [reply db/migrate/20200203105647_create_users.rb]: As the we are looking for the login method will require the users data.

File: public/docs/v1/style.css (hunk 9 above)
Comment (Devyani606) [reply public/docs/v1/style.css]: done

File: app/controllers/v1/sessions_controller.rb (hunk 5 above)
Comment (Devyani606) [reply app/controllers/v1/sessions_controller.rb]: done

File: app/controllers/v1/base_controller.rb [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+      end
+    end
+
+    def token_invalid
+      render_json(I18n.t("session.expired"), "less time", :unauthorized)
+    end
+
+    def switch_tenant!
+      @current_tenant = Organization.find_by(slug: @payload[:slug])
+      render_json("Tenant Not found", "Tenant doesn't exists", :unauthorized) unless @current_tenant
```
Comment (smriti154) [inline app/controllers/v1/base_controller.rb]: use localization

File: app/controllers/v1/base_controller.rb [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+      Apartment::Tenant.switch!(@current_tenant.slug)
+    end
+
+    def invalid_authentication
+      render_json("User Not Present", "Current user not found", :unauthorized)
```
Comment (smriti154) [inline app/controllers/v1/base_controller.rb]: use localization

File: app/controllers/v1/users_controller.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+module V1
+  class UsersController < ApplicationController
+    def index
+      @user = User.all
+      render json: @user
+    end
+  end
+end
```
Comment (smriti154) [inline app/controllers/v1/users_controller.rb]: this should not be present in this PR

**Review by smriti154 — APPROVED** [review]

**Pull request merged**